pkg/codegen/hcl2/rewrite_convert.go

```diff
@@ -100,37 +97,52 @@ func RewriteConversions(x model.Expression, to model.Type) model.Expression {
 			valueType, diags := to.Traverse(traverser)
 			contract.Ignore(diags)
 
-			item.Key = RewriteConversions(item.Key, model.InputType(model.StringType))
-			item.Value = RewriteConversions(item.Value, model.InputType(valueType.(model.Type)))
+			var valueChanged bool
+			item.Key, _ = rewriteConversions(item.Key, model.InputType(model.StringType))
```

Comment:
What is the principle here and elsewhere - when do we discard `typecheck` vs. when we don't?

Reply:
We discard `typecheck` when the rewrite can't affect the type of the containing expression. I'll add a note to that effect.